subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/IndexedNormalizedFileSnapshot.java

```diff
@@ -38,4 +40,9 @@ public String getAbsolutePath() {
     public int getIndex() {
         return index;
     }
+
+    @Override
+    protected CharSequence getNormalizedPathSequence() {
+        return CharBuffer.wrap(absolutePath, index, absolutePath.length());
+    }
```

Comment:
Why is this better than `absolutePath.substring(index)`?

Reply:
The latter allocates a new char[]

Reply:
Hm, I can't find anything about this. Can you point me to where this is documeted?

I based my original code on the sources that come with the JDK, here's the version for JDK 1.8.0_144:

```java
    /**
     * Returns a string that is a substring of this string. The
     * substring begins with the character at the specified index and
     * extends to the end of this string. <p>
     * Examples:
     * <blockquote><pre>
     * "unhappy".substring(2) returns "happy"
     * "Harbison".substring(3) returns "bison"
     * "emptiness".substring(9) returns "" (an empty string)
     * </pre></blockquote>
     *
     * @param      beginIndex   the beginning index, inclusive.
     * @return     the specified substring.
     * @exception  IndexOutOfBoundsException  if
     *             {@code beginIndex} is negative or larger than the
     *             length of this {@code String} object.
     */
    public String substring(int beginIndex) {
        if (beginIndex < 0) {
            throw new StringIndexOutOfBoundsException(beginIndex);
        }
        int subLen = value.length - beginIndex;
        if (subLen < 0) {
            throw new StringIndexOutOfBoundsException(subLen);
        }
        return (beginIndex == 0) ? this : new String(value, beginIndex, subLen);
    }
```

Reply:
Look into that constructor and you'll see the copy.

Reply:
Damn. Thanks!